File: src/app/services/FundsTransfer.service.ts
import { Injectable } from '@angular/core';
import { FormGroup,  FormBuilder,  Validators } from '@angular/forms';
import { HttpClient } from '@angular/common/http';
import { Observable, Subject } from 'rxjs';
import {FundsTransfer} from '../models/FundsTransfer';
import {AccountService} from '../services/Account.service';
import {ExternalAccountService} from '../services/ExternalAccount.service';
import {CustomerService} from '../services/Customer.service';
import {TransactionService} from '../services/Transaction.service';
import { HelperBaseService } from './helperbase.service';

@Injectable({
	providedIn: 'root'
})

export class FundsTransferService extends HelperBaseService {

	//********************************************************************
	// general holder 
	//********************************************************************
	fundsTransfer : FundsTransfer;

	//********************************************************************
	// Catch all for the return value of a service call
	//********************************************************************
	result: any;

	//********************************************************************
	// sole constructor, injected with the HttpClient
	//********************************************************************
	constructor(private http: HttpClient) {
		super();
	}

		//********************************************************************
	// add a FundsTransfer
	// returns the results untouched as a JSON representation
	// delegates via URI
	//********************************************************************
	addFundsTransfer(transferReference, amount, requestedDate, executionDate, purpose, feeAmount, SourceAccount, DestinationAccount, ExternalBeneficiary, InitiatedBy, Transactions, Method, Status) : Observable<any> {
		const uri_ = this.apiUrl + '/FundsTransfer/create';
		const obj = {
			      		transferReference: transferReference,
      		amount: amount,
      		requestedDate: requestedDate,
      		executionDate: executionDate,
      		purpose: purpose,
      		feeAmount: feeAmount,
      		SourceAccount: SourceAccount != null && SourceAccount.length > 0 ? SourceAccount : null,
      		DestinationAccount: DestinationAccount != null && DestinationAccount.length > 0 ? DestinationAccount : null,
      		ExternalBeneficiary: ExternalBeneficiary != null && ExternalBeneficiary.length > 0 ? ExternalBeneficiary : null,
      		InitiatedBy: InitiatedBy != null && InitiatedBy.length > 0 ? InitiatedBy : null,
      		Transactions: Transactions != null && Transactions.length > 0 ? Transactions : null,
      		Method: Method,
			Status: Status
		};

		return this.http.post(uri_, obj);
	}

	//********************************************************************
	// update a FundsTransfer
	// returns an Observable
	// delegates via URI
	//********************************************************************
		updateFundsTransfer(transferReference, amount, requestedDate, executionDate, purpose, feeAmount, SourceAccount, DestinationAccount, ExternalBeneficiary, InitiatedBy, Transactions, Method, Status, id)  :  Observable<any>  {
			const uri_ = this.apiUrl + '/FundsTransfer/update/' + id;
		const obj = {
				      		transferReference: transferReference,
      		amount: amount,
      		requestedDate: requestedDate,
      		executionDate: executionDate,
      		purpose: purpose,
      		feeAmount: feeAmount,
      		SourceAccount: SourceAccount != null && SourceAccount.length > 0 ? SourceAccount : null,
      		DestinationAccount: DestinationAccount != null && DestinationAccount.length > 0 ? DestinationAccount : null,
      		ExternalBeneficiary: ExternalBeneficiary != null && ExternalBeneficiary.length > 0 ? ExternalBeneficiary : null,
      		InitiatedBy: InitiatedBy != null && InitiatedBy.length > 0 ? InitiatedBy : null,
      		Transactions: Transactions != null && Transactions.length > 0 ? Transactions : null,
      		Method: Method,
			Status: Status
		};
		return this.http.post(uri_, obj);
	}

	//********************************************************************
	// delete a FundsTransfer
	// returns an Observable
	// delegates via URI
	//********************************************************************
	deleteFundsTransfer(id)  : Observable<any> {
		const uri_ = this.apiUrl + '/FundsTransfer/delete/' + id;

		return this.http.get(uri_);
	}
	
	//********************************************************************
	// loads a FundsTransfer
	// returns the results untouched as an Observable FundsTransfer
	// FundsTransfer model
	// delegates via URI
	//********************************************************************
	getFundsTransfer(id) : Observable<FundsTransfer> {
		const uri_ = this.apiUrl + '/FundsTransfer/load/' + id;

		return this.http.get<FundsTransfer>(uri_);
	}
	
	//********************************************************************
	// gets all FundsTransfer
	// returns the results untouched as JSON representation of an
	// Observable array of FundsTransfer models
	// delegates via URI
	//********************************************************************
	getFundsTransfers() : Observable<FundsTransfer[]> {
		const uri_ = this.apiUrl + '/FundsTransfer/';

		return this
			.http.get<FundsTransfer[]>(uri_);
	}
	
		
	//********************************************************************
	// assigns a SourceAccount on a FundsTransfer
	// returns an Observable
	// delegates via URI
	//********************************************************************
	assignSourceAccount( fundsTransferId, _sourceAccountId ): Observable<any> {

		// get the FundsTransfer from storage
		this.loadHelper( fundsTransferId );

	// get the Account from storage
	var tmp 	= new AccountService(this.http).getAccount(_sourceAccountId);

	// assign the SourceAccount
	this.fundsTransfer.sourceAccount = tmp;

	// save the FundsTransfer
	return this.saveHelper();
}

	//********************************************************************
	// unassigns a SourceAccount on a FundsTransfer
	// returns an Observable
	// delegates via URI
	//********************************************************************
	unassignSourceAccount( fundsTransferId ): Observable<any> {

		// get the FundsTransfer from storage
		this.loadHelper( fundsTransferId );

	// assign SourceAccount to null
	this.fundsTransfer.sourceAccount = null;

	// save the FundsTransfer
	return this.saveHelper();
}

	
	//********************************************************************
	// assigns a DestinationAccount on a FundsTransfer
	// returns an Observable
	// delegates via URI
	//********************************************************************
	assignDestinationAccount( fundsTransferId, _destinationAccountId ): Observable<any> {

		// get the FundsTransfer from storage
		this.loadHelper( fundsTransferId );

	// get the Account from storage
	var tmp 	= new AccountService(this.http).getAccount(_destinationAccountId);

	// assign the DestinationAccount
	this.fundsTransfer.destinationAccount = tmp;

	// save the FundsTransfer
	return this.saveHelper();
}

	//********************************************************************
	// unassigns a DestinationAccount on a FundsTransfer
	// returns an Observable
	// delegates via URI
	//********************************************************************
	unassignDestinationAccount( fundsTransferId ): Observable<any> {

		// get the FundsTransfer from storage
		this.loadHelper( fundsTransferId );

	// assign DestinationAccount to null
	this.fundsTransfer.destinationAccount = null;

	// save the FundsTransfer
	return this.saveHelper();
}

	
	//********************************************************************
	// assigns a ExternalBeneficiary on a FundsTransfer
	// returns an Observable
	// delegates via URI
	//********************************************************************
	assignExternalBeneficiary( fundsTransferId, _externalBeneficiaryId ): Observable<any> {

		// get the FundsTransfer from storage
		this.loadHelper( fundsTransferId );

	// get the ExternalAccount from storage
	var tmp 	= new ExternalAccountService(this.http).getExternalAccount(_externalBeneficiaryId);

	// assign the ExternalBeneficiary
	this.fundsTransfer.externalBeneficiary = tmp;

	// save the FundsTransfer
	return this.saveHelper();
}

	//********************************************************************
	// unassigns a ExternalBeneficiary on a FundsTransfer
	// returns an Observable
	// delegates via URI
	//********************************************************************
	unassignExternalBeneficiary( fundsTransferId ): Observable<any> {

		// get the FundsTransfer from storage
		this.loadHelper( fundsTransferId );

	// assign ExternalBeneficiary to null
	this.fundsTransfer.externalBeneficiary = null;

	// save the FundsTransfer
	return this.saveHelper();
}

	
	//********************************************************************
	// assigns a InitiatedBy on a FundsTransfer
	// returns an Observable
	// delegates via URI
	//********************************************************************
	assignInitiatedBy( fundsTransferId, _initiatedById ): Observable<any> {

		// get the FundsTransfer from storage
		this.loadHelper( fundsTransferId );

	// get the Customer from storage
	var tmp 	= new CustomerService(this.http).getCustomer(_initiatedById);

	// assign the InitiatedBy
	this.fundsTransfer.initiatedBy = tmp;

	// save the FundsTransfer
	return this.saveHelper();
}

	//********************************************************************
	// unassigns a InitiatedBy on a FundsTransfer
	// returns an Observable
	// delegates via URI
	//********************************************************************
	unassignInitiatedBy( fundsTransferId ): Observable<any> {

		// get the FundsTransfer from storage
		this.loadHelper( fundsTransferId );

	// assign InitiatedBy to null
	this.fundsTransfer.initiatedBy = null;

	// save the FundsTransfer
	return this.saveHelper();
}

	
		//********************************************************************
	// adds one or more transactionsIds as a Transactions
	// to a FundsTransfer
	// returns a Promise
	// delegates via URI
	//********************************************************************
	addTransactions( fundsTransferId, transactionsIds ): Observable<any> {

		// get the FundsTransfer
		this.loadHelper( fundsTransferId );

	// split on a comma with no spaces
	var idList = transactionsIds.split(',')

	// iterate over array of transactions ids
	idList.forEach(function (id) {
		// read the Transaction
		var transaction = new TransactionService(this.http).getTransaction(id);
		// add the Transaction if not already assigned
		if ( this.fundsTransfer.transactions.indexOf(transaction) == -1 )
		this.fundsTransfer.transactions.push(transaction);
	});

	// save it
	return this.saveHelper();
}

	//********************************************************************
	// removes one or more transactionsIds as a Transactions
	// from a FundsTransfer
	// returns a Promise
	// delegates via URI
	//********************************************************************
	removeTransactions( fundsTransferId, transactionsIds ): Observable<any> {

		// get the FundsTransfer
		this.loadHelper( fundsTransferId );


	// split on a comma with no spaces
	var idList 					= transactionsIds.split(',');
	var transactions 	= this.fundsTransfer.transactions;

	if ( transactions != null && transactionsIds != null ) {

		// iterate over array of transactions ids
		transactions.forEach(function (obj) {
			if ( transactionsIds.indexOf(obj._id) > -1 ) {
				// remove the Transaction
				this.fundsTransfer.transactions.pop(obj);
			}
		});

		// save it
		return this.saveHelper();
	}
}

	
	//********************************************************************
	// saveHelper - internal helper to save a FundsTransfer
	//********************************************************************
	saveHelper() : Observable<any> {

		const uri_ = this.apiUrl + '/FundsTransfer/update/' + this.fundsTransfer;

	return  this.http.post(uri_, this.fundsTransfer );
}

	//********************************************************************
	// loadHelper - internal helper to load a FundsTransfer
	//********************************************************************	
	loadHelper( id ) {
		this.getFundsTransfer(id)
			.subscribe((res : FundsTransfer) => {
				this.fundsTransfer = res;
			});
	}
}

File: bankingonaspdotnet/Service/FundsTransferService.cs

using bankingonaspdotnet.Domain;
using bankingonaspdotnet.Persistence;
using bankingonaspdotnet.Contracts;
using bankingonaspdotnet.Telemetry;

namespace bankingonaspdotnet.Service;

public interface IFundsTransferService
{

    Task Create(FundsTransfer model, CancellationToken cancellationToken);
    Task<bool> Update(FundsTransfer model, CancellationToken cancellationToken);
    Task<FundsTransfer?> Get(IdentifierRequest identifier, CancellationToken cancellationToken);
    Task<IReadOnlyList<FundsTransfer>> GetAll(CancellationToken cancellationToken);
    Task<bool> Delete(IdentifierRequest identifier, CancellationToken cancellationToken);
    // ------------------------------
    // Single Associations
    // -------------------------------
    Task<bool> AssignSourceAccount(AssociationRequest request, CancellationToken cancellationToken);
    Task<bool> UnassignSourceAccount(AssociationRequest request, CancellationToken cancellationToken);
    Task<bool> AssignDestinationAccount(AssociationRequest request, CancellationToken cancellationToken);
    Task<bool> UnassignDestinationAccount(AssociationRequest request, CancellationToken cancellationToken);
    Task<bool> AssignExternalBeneficiary(AssociationRequest request, CancellationToken cancellationToken);
    Task<bool> UnassignExternalBeneficiary(AssociationRequest request, CancellationToken cancellationToken);
    Task<bool> AssignInitiatedBy(AssociationRequest request, CancellationToken cancellationToken);
    Task<bool> UnassignInitiatedBy(AssociationRequest request, CancellationToken cancellationToken);

    Task<bool> AddToTransactions(MultipleAssociationRequest request, CancellationToken cancellationToken);
    Task<bool> RemoveFromTransactions(MultipleAssociationRequest request, CancellationToken cancellationToken);

}

public class FundsTransferService : IFundsTransferService
{
    private readonly ApplicationTelemetry _telemetry;
    private readonly IFundsTransferRepository _repository;
    private readonly ILogger<FundsTransferService> _logger;
    private readonly IServiceResolver _serviceResolver;


    public FundsTransferService(
        ApplicationTelemetry telemetry,
        IFundsTransferRepository repository,
        ILogger<FundsTransferService> logger,
        IServiceResolver serviceResolver)
    {
        _telemetry = telemetry;
        _repository = repository;
        _logger = logger;
        _serviceResolver = serviceResolver;
    }

    public async Task Create(FundsTransfer model, CancellationToken cancellationToken)
    {
        try
        {
            await _telemetry.Execute(
                "FundsTransfer",
                "CreateFundsTransfer",
                () => _repository.AddAsync(model, cancellationToken));
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
        }
    }

    public async Task<bool> Update(FundsTransfer model, CancellationToken cancellationToken)
    {
        try
        {
            var existing = await _repository.GetByIdAsync(model.Id, cancellationToken);
            if (existing is null)
            {
                return false;
            }
            existing.TransferReference = model.TransferReference;
            existing.Amount = model.Amount;
            existing.RequestedDate = model.RequestedDate;
            existing.ExecutionDate = model.ExecutionDate;
            existing.Purpose = model.Purpose;
            existing.FeeAmount = model.FeeAmount;
            existing.Method = model.Method;
            existing.Status = model.Status;

            await _telemetry.Execute(
                "FundsTransfer",
                "UpdateFundsTransfer",
                () => _repository.UpdateAsync(existing, cancellationToken));
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }

    public Task<FundsTransfer?> Get(IdentifierRequest identifier, CancellationToken cancellationToken)
    => _repository.GetByIdAsync(identifier.Id, cancellationToken);

    public Task<IReadOnlyList<FundsTransfer>> GetAll(CancellationToken cancellationToken)
    => _repository.GetAllAsync(cancellationToken);

    public async Task<bool> Delete(IdentifierRequest identifier, CancellationToken cancellationToken)
    {
        var existing = await _repository.GetByIdAsync(identifier.Id, cancellationToken);
        if (existing is null)
        {
            return false;
        }

        try
        {
            await _telemetry.Execute(
                "FundsTransfer",
                "UpdateFundsTransfer",
                () => _repository.DeleteAsync(existing, cancellationToken));
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }

    public async Task<bool> AssignSourceAccount(AssociationRequest request, CancellationToken cancellationToken)
    {

        var parent = await _repository.GetByIdAsync(request.ParentId, cancellationToken);
        if (parent is null)
        {
            _logger.LogError("No FundsTransfer found using Id {ParentId}", request.ParentId);
            return false;
        }

        try
        {
            var childRequest = new IdentifierRequest
            {
                Id = request.ChildId,
            };

            var child = await _serviceResolver.Get<AccountService>().Get(childRequest, cancellationToken);
            parent.SourceAccount = child;
            await Update(parent, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }

    public async Task<bool> UnassignSourceAccount(AssociationRequest request, CancellationToken cancellationToken)
    {
        var parent = await _repository.GetByIdAsync(request.ParentId, cancellationToken);
        if (parent is null)
        {
            _logger.LogError("No FundsTransfer found using Id {ParentId}", request.ParentId);
            return false;
        }

        try
        {
            parent.SourceAccount = null;
            await Update(parent, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }

    public async Task<bool> AssignDestinationAccount(AssociationRequest request, CancellationToken cancellationToken)
    {

        var parent = await _repository.GetByIdAsync(request.ParentId, cancellationToken);
        if (parent is null)
        {
            _logger.LogError("No FundsTransfer found using Id {ParentId}", request.ParentId);
            return false;
        }

        try
        {
            var childRequest = new IdentifierRequest
            {
                Id = request.ChildId,
            };

            var child = await _serviceResolver.Get<AccountService>().Get(childRequest, cancellationToken);
            parent.DestinationAccount = child;
            await Update(parent, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }

    public async Task<bool> UnassignDestinationAccount(AssociationRequest request, CancellationToken cancellationToken)
    {
        var parent = await _repository.GetByIdAsync(request.ParentId, cancellationToken);
        if (parent is null)
        {
            _logger.LogError("No FundsTransfer found using Id {ParentId}", request.ParentId);
            return false;
        }

        try
        {
            parent.DestinationAccount = null;
            await Update(parent, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }

    public async Task<bool> AssignExternalBeneficiary(AssociationRequest request, CancellationToken cancellationToken)
    {

        var parent = await _repository.GetByIdAsync(request.ParentId, cancellationToken);
        if (parent is null)
        {
            _logger.LogError("No FundsTransfer found using Id {ParentId}", request.ParentId);
            return false;
        }

        try
        {
            var childRequest = new IdentifierRequest
            {
                Id = request.ChildId,
            };

            var child = await _serviceResolver.Get<ExternalAccountService>().Get(childRequest, cancellationToken);
            parent.ExternalBeneficiary = child;
            await Update(parent, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }

    public async Task<bool> UnassignExternalBeneficiary(AssociationRequest request, CancellationToken cancellationToken)
    {
        var parent = await _repository.GetByIdAsync(request.ParentId, cancellationToken);
        if (parent is null)
        {
            _logger.LogError("No FundsTransfer found using Id {ParentId}", request.ParentId);
            return false;
        }

        try
        {
            parent.ExternalBeneficiary = null;
            await Update(parent, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }

    public async Task<bool> AssignInitiatedBy(AssociationRequest request, CancellationToken cancellationToken)
    {

        var parent = await _repository.GetByIdAsync(request.ParentId, cancellationToken);
        if (parent is null)
        {
            _logger.LogError("No FundsTransfer found using Id {ParentId}", request.ParentId);
            return false;
        }

        try
        {
            var childRequest = new IdentifierRequest
            {
                Id = request.ChildId,
            };

            var child = await _serviceResolver.Get<CustomerService>().Get(childRequest, cancellationToken);
            parent.InitiatedBy = child;
            await Update(parent, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }

    public async Task<bool> UnassignInitiatedBy(AssociationRequest request, CancellationToken cancellationToken)
    {
        var parent = await _repository.GetByIdAsync(request.ParentId, cancellationToken);
        if (parent is null)
        {
            _logger.LogError("No FundsTransfer found using Id {ParentId}", request.ParentId);
            return false;
        }

        try
        {
            parent.InitiatedBy = null;
            await Update(parent, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }


    public async Task<bool> AddToTransactions(MultipleAssociationRequest request, CancellationToken cancellationToken)
    {
        try
        {
            await _telemetry.Execute(
                "FundsTransfer",
                "AddToTransactions",
                () => _repository.AddToTransactionsAsync(request, cancellationToken));
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }

    public async Task<bool> RemoveFromTransactions(MultipleAssociationRequest request, CancellationToken cancellationToken)
    {
        try
        {
            await _telemetry.Execute(
                "FundsTransfer",
                "RemoveFromTransactions",
                () => _repository.RemoveFromTransactionsAsync(request, cancellationToken));
        }
        catch (Exception ex)
        {
            _logger.LogError(
                    ex,
                    "Unexpected error while creating Transaction.");
            return false;
        }
        return true;
    }



}
